### PR TITLE
Add Submodule: Fill combobox with branches regardless of the repo URL type

### LINKED ISSUE #6556: Add Submodule: Fill combobox with branches regardless of the repo URL's type
## Current behaviour

I have tried to add a (first) submodule to a repo.
The exception is thrown when I try to open the yet empty `Branch` combobox.

This combobox is not filled for remote repos at all, regardless whether specified with HTTPS URL or with SSH URL.

## Expected behaviour

list of available branches filled correctly
no unhandled exception
no exception for a valid URL

## Steps to reproduce

- create a standalone clone of the repo which shall be used a submodule
- copy its `URL` from the `Remote repositories` dialog ("git@host:user/path.git")
- open GE for the destination repo
- `Manage submodules`, `Add submodule`
- paste the URL
- drop down the `Branch` combobox, which was empty yet

## Error details 3.2.0
- combobox does not contain the branches of the sub-repo unless an URI of type `//file:` is given

(no unhandled exception)

## Error Details 3.1.0
```
System.UriFormatException: Ungültiger URI: Das URI-Schema ist ungültig.
   bei System.Uri.CreateThis(String uri, Boolean dontEscape, UriKind uriKind)
   bei GitCommands.PathUtil.NormalizePath(String path)
   bei GitCommands.GitModule..ctor(String workingDir)
   bei GitUI.CommandsDialogs.SubmodulesDialog.FormAddSubmodule.BranchDropDown(Object sender, EventArgs e)
   bei System.Windows.Forms.ComboBox.OnDropDown(EventArgs e)
   bei System.Windows.Forms.ComboBox.WmReflectCommand(Message& m)
   bei System.Windows.Forms.ComboBox.WndProc(Message& m)
   bei System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)
```

## Environment
- Git Extensions 3.1.0
- Build 678b69e632a2bff81c5e093770f605e0fe1151c8
- Git 2.21.0.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.3324.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by mstv**
Fixes #6556

## Proposed changes

- use "git ls-remote" to get possible submodule branches

### Before

grafik

### After

grafik

## Test methodology 

- manual testing
- add a NUnit test

## Test environment(s) 

- Git Extensions 3.3.0
- Build db1c36b00e745692448ef8b877dd0b3b858ad3b6
- Git 2.23.0.windows.1
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4018.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs [hunk 1]
```diff
@@ -70,20 +72,19 @@ private void DirectorySelectedIndexChanged(object sender, EventArgs e)
...
-
-            if (module.IsValidGitWorkingDir())
-            {
-                heads.AddRange(module.GetRefs(false));
-            }
-
-            Branch.DisplayMember = nameof(IGitRef.Name);
-            Branch.DataSource = heads;
+            var gitExecutable = new Executable(() => AppSettings.GitCommand);
+            var heads = gitExecutable.GetOutput(GitCommandHelpers.LsRemotesCmd(url: Directory.Text));
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs]: Why is the command executed outside the module context? I.e. not via GitModule instance?

I think a correct approach would be to construct a module here:
```cs
var submodule = new GitModule(Directory.Text);
var heads = submodule.GitExecutable.GetOutput(GitCommandHelpers.LsRemotesCmd(submodule.WorkingDirectory));
```
or even better encapsulate the functionality int `GitModule`, so we have here:
```cs
var submodule = new GitModule(Directory.Text);
var heads = submodule.GetLsRemotes();
```

File: GitCommands/Git/GitCommandHelpers.cs [hunk 2]
```diff
@@ -1025,6 +1025,14 @@ public static ArgumentString FindRenamesAndCopiesOpts()
             return FindRenamesOpt() + findCopies;
         }
 
+        public static ArgumentString LsRemotesCmd(string url)
```
Comment (RussKie) [inline GitCommands/Git/GitCommandHelpers.cs]: This has a single use, I think we can inline it at the callsite until we have more uses for it.
Besides now we need to all sort of validations here, which makes it more complex than required

File: GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs (hunk 1 above)
Comment (mstv) [reply GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs]: That's why it worked for local directories only but not for SSH and HTTP[S] links.
`git ls-remotes <url>` is completely independent from a local repo clone.

File: GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs [hunk 3]
```diff
@@ -70,20 +72,19 @@ private void DirectorySelectedIndexChanged(object sender, EventArgs e)
...
+            var branches
+                = heads
+                    .SplitLines()
```
[the remark this answers is not in the document]
Comment (mstv) [reply GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs]: Shall I indent with 4 spaces or shall I align the dots?
```
var branches = heads.Splitlines()
    .Where(...)
```
or
```
var branches = heads.Splitlines()
                    .Where(...)
```

File: GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs [hunk 4]
```diff
@@ -70,20 +73,19 @@ private void DirectorySelectedIndexChanged(object sender, EventArgs e)
...
-
-            if (module.IsValidGitWorkingDir())
-            {
-                heads.AddRange(module.GetRefs(false));
-            }
-
-            Branch.DisplayMember = nameof(IGitRef.Name);
-            Branch.DataSource = heads;
+            var url = Directory.Text;
+            var gitArguments = new GitArgumentBuilder("ls-remote") { url.ToPosixPath().Quote() };
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs]: Should have --heads, no need to fetch tags too

Comment (mstv) [reply GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs]: > Should have --heads, no need to fetch tags too

Good point, thank you.
I've kept the check for valid branch refs. On error, the drop down list will be empty.

File: GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/SubmodulesDialog/FormAddSubmodule.cs]: Right, thank you.
It would be good to add comments explaining this.

**Comment by RussKie** [comment]
Can we restructure the change a little and make it testable?

```cs
        private void BranchDropDown(object sender, EventArgs e)
        {
            var gitExecutable = new Executable(() => AppSettings.GitCommand);
            Branch.DataSource = LoadBranches(gitExecutable, Directory.Text);
        }

        private <type> LoadRemoteRepoRefs(IExecutable gitExecutable, string url)
        {
			if (string.IsNullOrWhiteSpace(url))
			{
				return Array.Empty<string>(); // or whatever else appropriate
			}

            var gitArguments = new GitArgumentBuilder("ls-remote") { "--heads", url.ToPosixPath().Quote() };
            var heads = gitExecutable.GetOutput(gitArguments);
            return heads.SplitLines()
                                .Select(head =>
                                {
                                    int branchIndex = head.IndexOf(GitRefName.RefsHeadsPrefix);
                                    return branchIndex == -1 ? null : head.Substring(branchIndex + GitRefName.RefsHeadsPrefix.Length);
                                })
                                .Where(branch => branch != null)
                                .ToImmutableList();
        }

```

**Pull request merged**